### DISCUSSION TITLE
How to copy an object from one db to another using orm

### DISCUSSION THREAD

**Discussion by taranlu-houzz** [sqlalchemy/sqlalchemy#7435] [Usage Questions] [answered]
### Describe the bug

I am using `sqlmodel`, but I believe this is actually due to underlaying behavior in `sqlalchemy`.

Basically, I have a model where the timestamp columns are set to use `server_default=sqlalchemy.func.now()`, which works great. However, if I happen to try to add a model the has values set for those columns, the add silently fails.

My use case is that I am trying to copy the data in a table from one engine to another to setup/cleanup fixtures for testing. In my case, this means copying the data from a `sqlite` db file to an in-memory `sqlite` db that a test client is connected to.

### To Reproduce

```python
"""server_default issue"""


from datetime import datetime
from typing import Optional

import sqlalchemy
from sqlalchemy import func
from sqlmodel import (
    Field,
    SQLModel,
    Session,
    select,
)

try:
    from rich import print
except ImportError:
    pass


class TestModel(SQLModel, table=True):

    product_id: int = Field(primary_key=True)
    name: str
    created_time: Optional[datetime] = Field(
        default=None,
        sa_column_kwargs={
            "server_default": func.now(),
        },
    )
    updated_time: Optional[datetime] = Field(
        default=None,
        sa_column_kwargs={
            "server_default": func.now(),
        },
    )


# Setup
engine = sqlalchemy.create_engine(
    "sqlite://",
    future=True,
    connect_args={"check_same_thread": False},
    echo=True,
)
SQLModel.metadata.create_all(engine)


session = Session(engine)


# Able to add item as long as server_defaul columns are None
new_item = TestModel(product_id=15, name="A")
session.add(new_item)
session.commit()

# Unable to add item if have values set
another_item = TestModel(product_id=99, name="B", created_time=datetime.now())
session.commit()

# Show data in db
data = session.exec(select(TestModel)).all()
print(data)
```

### Error

This issue does not cause any error.

### Versions

- OS: macOS 10.15.7
- Python: 3.10.0
- SQLAlchemy: 1.4.27
- Database: sqlite
- DBAPI (eg: psycopg, cx_oracle, mysqlclient): N/A?

### Additional context

```sh
❯ python sqlalchemy_issue.py
2021-12-10 03:13:31,003 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2021-12-10 03:13:31,003 INFO sqlalchemy.engine.Engine PRAGMA main.table_info("testmodel")
2021-12-10 03:13:31,003 INFO sqlalchemy.engine.Engine [raw sql] ()
2021-12-10 03:13:31,003 INFO sqlalchemy.engine.Engine PRAGMA temp.table_info("testmodel")
2021-12-10 03:13:31,004 INFO sqlalchemy.engine.Engine [raw sql] ()
2021-12-10 03:13:31,004 INFO sqlalchemy.engine.Engine
CREATE TABLE testmodel (
        product_id INTEGER NOT NULL,
        name VARCHAR NOT NULL,
        created_time DATETIME DEFAULT (CURRENT_TIMESTAMP),
        updated_time DATETIME DEFAULT (CURRENT_TIMESTAMP),
        PRIMARY KEY (product_id)
)


2021-12-10 03:13:31,004 INFO sqlalchemy.engine.Engine [no key 0.00006s] ()
2021-12-10 03:13:31,004 INFO sqlalchemy.engine.Engine CREATE INDEX ix_testmodel_updated_time ON testmodel (updated_time)
2021-12-10 03:13:31,004 INFO sqlalchemy.engine.Engine [no key 0.00006s] ()
2021-12-10 03:13:31,005 INFO sqlalchemy.engine.Engine CREATE INDEX ix_testmodel_product_id ON testmodel (product_id)
2021-12-10 03:13:31,005 INFO sqlalchemy.engine.Engine [no key 0.00005s] ()
2021-12-10 03:13:31,005 INFO sqlalchemy.engine.Engine CREATE INDEX ix_testmodel_created_time ON testmodel (created_time)
2021-12-10 03:13:31,005 INFO sqlalchemy.engine.Engine [no key 0.00006s] ()
2021-12-10 03:13:31,005 INFO sqlalchemy.engine.Engine CREATE INDEX ix_testmodel_name ON testmodel (name)
2021-12-10 03:13:31,005 INFO sqlalchemy.engine.Engine [no key 0.00005s] ()
2021-12-10 03:13:31,005 INFO sqlalchemy.engine.Engine COMMIT
2021-12-10 03:13:31,006 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2021-12-10 03:13:31,007 INFO sqlalchemy.engine.Engine INSERT INTO testmodel (product_id, name) VALUES (?, ?)
2021-12-10 03:13:31,007 INFO sqlalchemy.engine.Engine [generated in 0.00010s] (15, 'A')
2021-12-10 03:13:31,007 INFO sqlalchemy.engine.Engine COMMIT
2021-12-10 03:13:31,008 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2021-12-10 03:13:31,008 INFO sqlalchemy.engine.Engine SELECT testmodel.product_id, testmodel.name, testmodel.created_time, testmodel.updated_time
FROM testmodel
2021-12-10 03:13:31,008 INFO sqlalchemy.engine.Engine [no key 0.00007s] ()
[TestModel(created_time=datetime.datetime(2021, 12, 10, 11, 13, 31), updated_time=datetime.datetime(2021, 12, 10, 11, 13, 31), product_id=15, name='A')]
```

**Comment by zzzeek** [comment]
hi there -

we don't support SQLModel here however your test appears to have a simple mistake of not adding the second object to the session:

```py
# Unable to add item if have values set
another_item = TestModel(product_id=99, name="B", created_time=datetime.now())
session.commit()
```


you would need to do this so that the object is inserted

```py
another_item = TestModel(product_id=99, name="B", created_time=datetime.now())
session.add(another_item)
session.commit()
```

**Reply by taranlu-houzz** [reply]
@zzzeek Ah, shoot. Yes fixing that simple mistake does make the test code above work correctly.

This is just a cut down example of an issue I am still running into though, so I will try to investigate further in my actual code to see if I can reproduce behavior.

**Reply by taranlu-houzz** [reply]
@zzzeek So, with this updated test code (which is closer to what my actual code entails), I am able to reproduce the issue:
```python
"""server_default issue"""


from datetime import datetime
from typing import Optional

from sqlalchemy import func
from sqlmodel import (
    Field,
    SQLModel,
    Session,
    create_engine,
    select,
)

try:
    from rich import print
except ImportError:
    pass


class Product(SQLModel, table=True):

    product_id: int = Field(primary_key=True)
    created_time: Optional[datetime] = Field(
        default=None,
        sa_column_kwargs={
            "server_default": func.now(),
        },
    )
    updated_time: Optional[datetime] = Field(
        default=None,
        sa_column_kwargs={
            "server_default": func.now(),
        },
    )


# Setup sqlite file db
engine0 = create_engine(
    "sqlite:///sqlalchemy_issue.db",
    future=True,
    connect_args={"check_same_thread": False},
    echo=True,
)
SQLModel.metadata.create_all(engine0)

# Setup in-memory db
engine1 = create_engine(
    "sqlite://",
    future=True,
    connect_args={"check_same_thread": False},
    echo=True,
)
SQLModel.metadata.create_all(engine1)

# Rebuild sqlite file db if it does not exist
with Session(engine0) as session:
    data = session.exec(select(Product)).all()
    if not data:
        print("")
        print("--> Build engine0 db sqlite file: START")
        for pid in range(6):
            session.add(Product(product_id=pid))
        session.commit()
        print("Added data to engine0")
        print("<-- Build engine0 db sqlite file: END")

# Get items from engine0 db
print("")
print("--> Get data from engine0: START")
with Session(engine0) as session:
    data0 = session.exec(select(Product)).all()
    print("Got data from engine0")
    print(data0)
print("<-- Get data from engine0: END")

# Add items to engine1 db
print("")
print("--> Add data to engine1: START")
with Session(engine1) as session:
    for item in data0:
        session.add(item)
    session.commit()
    print("Added data to engine1")
print("<-- Add data to engine1: END")

# Get items from engine1 db
print("")
print("--> Get data from engine1: START")
with Session(engine1) as session:
    data1 = session.exec(select(Product)).all()
    print("Got data from engine1")
    print(data1)
print("<-- Get data from engine1: END")

# Add item without server_default columns
print("")
print("--> Add new_item to engine1: START")
with Session(engine1) as session:
    new_item = Product(product_id=99)
    session.add(new_item)
    session.commit()
    print("Added new_item to engine1")
print("<-- Add data to engine1: END")

# Get items from engine1 db (again)
print("")
print("--> Get data from engine1 (again): START")
with Session(engine1) as session:
    data1 = session.exec(select(Product)).all()
    print("Got data from engine1")
    print(data1)
print("<-- Get data from engine1 (again): END")
```

The output from running the above gives:
```sh
2021-12-10 11:50:41,014 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2021-12-10 11:50:41,014 INFO sqlalchemy.engine.Engine PRAGMA main.table_info("product")
2021-12-10 11:50:41,014 INFO sqlalchemy.engine.Engine [raw sql] ()
2021-12-10 11:50:41,014 INFO sqlalchemy.engine.Engine PRAGMA temp.table_info("product")
2021-12-10 11:50:41,014 INFO sqlalchemy.engine.Engine [raw sql] ()
2021-12-10 11:50:41,015 INFO sqlalchemy.engine.Engine 
CREATE TABLE product (
	product_id INTEGER NOT NULL, 
	created_time DATETIME DEFAULT (CURRENT_TIMESTAMP), 
	updated_time DATETIME DEFAULT (CURRENT_TIMESTAMP), 
	PRIMARY KEY (product_id)
)


2021-12-10 11:50:41,015 INFO sqlalchemy.engine.Engine [no key 0.00006s] ()
2021-12-10 11:50:41,016 INFO sqlalchemy.engine.Engine CREATE INDEX ix_product_product_id ON product (product_id)
2021-12-10 11:50:41,016 INFO sqlalchemy.engine.Engine [no key 0.00006s] ()
2021-12-10 11:50:41,017 INFO sqlalchemy.engine.Engine CREATE INDEX ix_product_updated_time ON product (updated_time)
2021-12-10 11:50:41,017 INFO sqlalchemy.engine.Engine [no key 0.00006s] ()
2021-12-10 11:50:41,018 INFO sqlalchemy.engine.Engine CREATE INDEX ix_product_created_time ON product (created_time)
2021-12-10 11:50:41,018 INFO sqlalchemy.engine.Engine [no key 0.00006s] ()
2021-12-10 11:50:41,019 INFO sqlalchemy.engine.Engine COMMIT
2021-12-10 11:50:41,019 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2021-12-10 11:50:41,020 INFO sqlalchemy.engine.Engine PRAGMA main.table_info("product")
2021-12-10 11:50:41,020 INFO sqlalchemy.engine.Engine [raw sql] ()
2021-12-10 11:50:41,020 INFO sqlalchemy.engine.Engine PRAGMA temp.table_info("product")
2021-12-10 11:50:41,020 INFO sqlalchemy.engine.Engine [raw sql] ()
2021-12-10 11:50:41,020 INFO sqlalchemy.engine.Engine 
CREATE TABLE product (
	product_id INTEGER NOT NULL, 
	created_time DATETIME DEFAULT (CURRENT_TIMESTAMP), 
	updated_time DATETIME DEFAULT (CURRENT_TIMESTAMP), 
	PRIMARY KEY (product_id)
)


2021-12-10 11:50:41,020 INFO sqlalchemy.engine.Engine [no key 0.00005s] ()
2021-12-10 11:50:41,020 INFO sqlalchemy.engine.Engine CREATE INDEX ix_product_product_id ON product (product_id)
2021-12-10 11:50:41,020 INFO sqlalchemy.engine.Engine [no key 0.00005s] ()
2021-12-10 11:50:41,020 INFO sqlalchemy.engine.Engine CREATE INDEX ix_product_updated_time ON product (updated_time)
2021-12-10 11:50:41,020 INFO sqlalchemy.engine.Engine [no key 0.00005s] ()
2021-12-10 11:50:41,021 INFO sqlalchemy.engine.Engine CREATE INDEX ix_product_created_time ON product (created_time)
2021-12-10 11:50:41,021 INFO sqlalchemy.engine.Engine [no key 0.00005s] ()
2021-12-10 11:50:41,021 INFO sqlalchemy.engine.Engine COMMIT
2021-12-10 11:50:41,021 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2021-12-10 11:50:41,022 INFO sqlalchemy.engine.Engine SELECT product.product_id, product.created_time, product.updated_time 
FROM product
2021-12-10 11:50:41,022 INFO sqlalchemy.engine.Engine [no key 0.00008s] ()

--> Build engine0 db sqlite file: START
2021-12-10 11:50:41,072 INFO sqlalchemy.engine.Engine INSERT INTO product (product_id) VALUES (?)
2021-12-10 11:50:41,072 INFO sqlalchemy.engine.Engine [generated in 0.00015s] ((0,), (1,), (2,), (3,), (4,), (5,))
2021-12-10 11:50:41,073 INFO sqlalchemy.engine.Engine COMMIT
Added data to engine0
<-- Build engine0 db sqlite file: END

--> Get data from engine0: START
2021-12-10 11:50:41,076 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2021-12-10 11:50:41,076 INFO sqlalchemy.engine.Engine SELECT product.product_id, product.created_time, product.updated_time 
FROM product
2021-12-10 11:50:41,076 INFO sqlalchemy.engine.Engine [no key 0.00009s] ()
Got data from engine0
[
    Product(updated_time=datetime.datetime(2021, 12, 10, 19, 50, 41), product_id=0, created_time=datetime.datetime(2021, 12, 
10, 19, 50, 41)),
    Product(updated_time=datetime.datetime(2021, 12, 10, 19, 50, 41), product_id=1, created_time=datetime.datetime(2021, 12, 
10, 19, 50, 41)),
    Product(updated_time=datetime.datetime(2021, 12, 10, 19, 50, 41), product_id=2, created_time=datetime.datetime(2021, 12, 
10, 19, 50, 41)),
    Product(updated_time=datetime.datetime(2021, 12, 10, 19, 50, 41), product_id=3, created_time=datetime.datetime(2021, 12, 
10, 19, 50, 41)),
    Product(updated_time=datetime.datetime(2021, 12, 10, 19, 50, 41), product_id=4, created_time=datetime.datetime(2021, 12, 
10, 19, 50, 41)),
    Product(updated_time=datetime.datetime(2021, 12, 10, 19, 50, 41), product_id=5, created_time=datetime.datetime(2021, 12, 
10, 19, 50, 41))
]
2021-12-10 11:50:41,082 INFO sqlalchemy.engine.Engine ROLLBACK
<-- Get data from engine0: END

--> Add data to engine1: START
Added data to engine1
<-- Add data to engine1: END

--> Get data from engine1: START
2021-12-10 11:50:41,084 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2021-12-10 11:50:41,084 INFO sqlalchemy.engine.Engine SELECT product.product_id, product.created_time, product.updated_time 
FROM product
2021-12-10 11:50:41,084 INFO sqlalchemy.engine.Engine [no key 0.00008s] ()
Got data from engine1
[]
2021-12-10 11:50:41,085 INFO sqlalchemy.engine.Engine ROLLBACK
<-- Get data from engine1: END

--> Add new_item to engine1: START
2021-12-10 11:50:41,086 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2021-12-10 11:50:41,087 INFO sqlalchemy.engine.Engine INSERT INTO product (product_id) VALUES (?)
2021-12-10 11:50:41,087 INFO sqlalchemy.engine.Engine [generated in 0.00010s] (99,)
2021-12-10 11:50:41,087 INFO sqlalchemy.engine.Engine COMMIT
Added new_item to engine1
<-- Add data to engine1: END

--> Get data from engine1 (again): START
2021-12-10 11:50:41,088 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2021-12-10 11:50:41,089 INFO sqlalchemy.engine.Engine SELECT product.product_id, product.created_time, product.updated_time 
FROM product
2021-12-10 11:50:41,089 INFO sqlalchemy.engine.Engine [no key 0.00008s] ()
Got data from engine1
[
    Product(updated_time=datetime.datetime(2021, 12, 10, 19, 50, 41), product_id=99, created_time=datetime.datetime(2021, 12, 
10, 19, 50, 41))
]
2021-12-10 11:50:41,090 INFO sqlalchemy.engine.Engine ROLLBACK
<-- Get data from engine1 (again): END
```

I am actually not so certain this is an issue with `sqlalchemy` anymore, so I guess maybe I should make an issue on the `sqlmodel` repo as well?

*Edit: Modified the test code to also generate the local `sqlite` db if it does not exist.

**Reply by CaselIT** [reply]
Not sure why the server default were mentioned here, since they have no influence in the above behaviour

The issue comes from trying to add objects that were obtained from one session on a db to another session attached to another db.
The objects obtained from a session are in persisted state, they move to detached once the original session is closed and when they are added to a new session they are assumed to be in the db since they have the primary key populated.
Replacing `session.add(item)` with `session.merge(item)` will work as you expect.
See also https://docs.sqlalchemy.org/en/14/orm/session_state_management.html#session-object-states

I guess this use case could be improved, but I would consider this an edge case

**Reply by taranlu-houzz** [reply]
@CaselIT Ah, thank you! That is exactly what I needed and it resolves my issue nicely. I thought that it was related to `server_default` since leaving that blank added the new entry the way I had expected.

**Reply by CaselIT** [reply]
@zzzeek do you think this use case is something that could be better supported without much effort?

**Reply by zzzeek** [reply]
"The issue comes from trying to add objects that were obtained from one session on a db to another session attached to another db." that use case?  yes you use either merge() or make_transient()

**Reply by zzzeek** [reply]
or persistent_to_detached().  it depends highly on whether or not the rows are present in the target database

**Reply by taranlu-houzz** [reply]
I am fairly new to this, but it would seem to me a decent way of testing an api with backend db connections would be to store several `sqlite` database files as part of the test suite data within a given project repo. Then, when running tests, all data from the `sqlite` db files would be loaded into in-memory `sqlite` databases which the test api client would connect to. This should allow the tests to run as quickly as possible for `sqlite` and would also make it easy to "reset" the test db to a known state. The db files could also be named, which would help with organization.

Basically, it would be nice to a function that can take an engine connected to a file based sqlite db and then clone data from it to another engine connected to an in-memory db.

**Reply by CaselIT** [reply]
> "The issue comes from trying to add objects that were obtained from one session on a db to another session attached to another db." that use case? yes you use either merge() or make_transient()

yep.

I was thinking about keeping in the instances from which db (url?) they were detached from. But it's probably not something it's worth doing

**Reply by CaselIT** [reply]
@taranlu-houzz it's probably faster to just make a copy of the sqlite db file for the current test than to use python do copy everything from one db to another.

Also I'm not sure if that test architecture would be scalable long term, since every time you change the schema you would need to (manually?) update a lot of dbs

**Reply by taranlu-houzz** [reply]
@CaselIT Ah, okay, good to know. In my case, it probably doesn't matter, since my test db has a tiny amount of data in it (probably premature optimization on my part). So, your suggestion would be to just copy the db file to a temp location, create a new engine/test client, use that for the test, then discard and make new ones for subsequent tests?

**Reply by CaselIT** [reply]
If you want to use template dbs, then yes.

As mentioned above, not sure if that's scalable long term

**Reply by taranlu-houzz** [reply]
Okay, I'll keep that in mind. It should be fine for my use case. Thank you again for your help @CaselIT @zzzeek!

**Comment by taranlu-houzz** [comment, marked as answer]
Using `session.merge()` as suggested above by @CaselIT solved my issue: https://github.com/sqlalchemy/sqlalchemy/discussions/7435#discussioncomment-1788215
